sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/query/DocumentQueryExecutionContextFactory.java

```diff
@@ -78,49 +84,105 @@
-                    .map(pkRanges -> Pair.of(
-                        pkRanges,
-                        partitionedQueryExecutionInfo.getQueryInfo()));
+
+                return getTargetRangesFromQueryPlan(cosmosQueryRequestOptions, collection, queryExecutionContext,
+                                                    partitionedQueryExecutionInfo, startTime, endTime);
             });
     }
 
+    private static <T extends Resource> Mono<Pair<List<PartitionKeyRange>, QueryInfo>> getTargetRangesFromQueryPlan(
+        CosmosQueryRequestOptions cosmosQueryRequestOptions, DocumentCollection collection,
+        DefaultDocumentQueryExecutionContext<T> queryExecutionContext,
+        PartitionedQueryExecutionInfo partitionedQueryExecutionInfo, Instant planFetchStartTime,
+        Instant planFetchEndTime) {
+        QueryInfo queryInfo =
+            partitionedQueryExecutionInfo.getQueryInfo();
+        queryInfo.setQueryPlanDiagnosticsContext(new QueryInfo.QueryPlanDiagnosticsContext(planFetchStartTime,
+                                                                                           planFetchEndTime));
+        List<Range<String>> queryRanges =
+            partitionedQueryExecutionInfo.getQueryRanges();
+
+        if (isScopedToSinglePartition(cosmosQueryRequestOptions)) {
+            PartitionKeyInternal internalPartitionKey =
+                BridgeInternal.getPartitionKeyInternal(cosmosQueryRequestOptions.getPartitionKey());
+            Range<String> range = Range
+                                      .getPointRange(internalPartitionKey
+                                                         .getEffectivePartitionKeyString(internalPartitionKey,
+                                                                                         collection
+                                                                                             .getPartitionKey()));
+            queryRanges = Collections.singletonList(range);
+        }
+        return
+            queryExecutionContext.getTargetPartitionKeyRanges(collection.getResourceId(), queryRanges)
+                .map(pkRanges -> Pair.of(
+                    pkRanges,
+                    partitionedQueryExecutionInfo.getQueryInfo()));
+    }
+
+    private static void tryCacheQueryPlan(
+        SqlQuerySpec query,
+        PartitionedQueryExecutionInfo partitionedQueryExecutionInfo,
+        LRUCache<String, PartitionedQueryExecutionInfo> queryPlanCache) {
+        QueryInfo queryInfo = partitionedQueryExecutionInfo.getQueryInfo();
+        if (canCacheQuery(queryInfo) && !queryPlanCache.containsKey(query.getQueryText())) {
+            try {
+                queryPlanCache.put(query.getQueryText(), partitionedQueryExecutionInfo);
+            } catch (ConcurrentModificationException exception) {
+                logger.error("Error caching query plan: ", exception);
```

Comment:
I don't think this is the current pattern. We shouldn't be using a cache which is not thread-safe.

Reply:
Agree. But this, at this point, is a experimental feature and adding strong thread safety is significantly reducing the perf of the cache. As the feature matures, we can even go with a unbounded ConcurrentMap

Reply:
Moved to a thread safe map for now without LRU.

Reply:
could you please remove this then?